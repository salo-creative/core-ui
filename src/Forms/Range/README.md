# Range

This is an input field with type=range, styled according to the theme. It can take most properties that an Input element can take, as well as a few others. The unique properties include:

* min. Number. The minimum value of the range. Default 0.
* max. Number. The maximum value of the range. Default 100.
* step. Number. The amount to increment by when the slider is manipulated.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Range from '@salo/core-ui/Forms/Range';
```

Implement as follows

```javascript
<Range
  error={ error }
  errorMessage={ errorMessage }
  helperText={ helperText }
  label={ label }
  name='story'
  min={ 0 }
  max={ 100 }
  value={ state.value }
  onChange={ ({ value }) => {
    store.set({ value });
  } }
  step={ step }
/>
```