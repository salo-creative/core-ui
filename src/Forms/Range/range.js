import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import Label from '../components/Label';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import {
  RangeSliderOuterWrapper,
  RangeSliderInnerWrapper,
  Slider,
  ValueDisplay,
  ValWrapper
} from './range.styles';


const Range = (props) => {
  const {
    disabled,
    error,
    errorMessage,
    helperText,
    label,
    margin,
    max,
    min,
    name,
    onBlur,
    onChange,
    required,
    size,
    step,
    value
  } = props;

  const calculateHeight = () => {
    switch (size) {
      case 'M':
        return '4rem';
      case 'L':
      default:
        return '4.5rem';
    }
  };

  const height = calculateHeight();

  const renderSlider = () => {
    if (min < max) {
      return (
        <React.Fragment>
          <Slider
            aria-required={ required.toString() }
            aria-invalid={ error.toString() }
            disabled={ disabled }
            type='range'
            id={ name }
            name={ name }
            min={ min }
            max={ max }
            value={ value }
            onBlur={ (e) => onBlur({ e, value: e.target.value }) }
            onChange={ (e) => onChange({ e, value: e.target.value }) }
            step={ step }
          />
          <ValueDisplay size={ height }>
            { value }
          </ValueDisplay>
        </React.Fragment>
      );
    }

    return 'The min value cannot be higher than the max.';
  };

  return (
    <RangeSliderOuterWrapper
      margin={ margin }
    >
      <Label
        error={ error }
        label={ label }
        name={ name }
        required={ required }
        size={ size }
      />
      <RangeSliderInnerWrapper
        height={ height }
      >
        { renderSlider() }
      </RangeSliderInnerWrapper>
      <ErrorText
        disabled={ disabled }
        error={ error }
        errorMessage={ errorMessage }
        size={ size }
      />
      <HelperText
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        size={ size }
      />
    </RangeSliderOuterWrapper>
  );
};

Range.defaultProps = {
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  helperText: '',
  label: '',
  margin: '0 0 2rem',
  onBlur: () => null,
  onChange: () => null,
  required: false,
  step: PropTypes.number,
  size: 'M',
  value: '',
  min: 0,
  max: 100
};

Range.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  helperText: PropTypes.string,
  label: PropTypes.string,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  step: PropTypes.number,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  min: PropTypes.number,
  max: PropTypes.number
};

export default Range;