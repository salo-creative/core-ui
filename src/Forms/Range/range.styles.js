import styled, { css } from 'styled-components';

export const RangeSliderOuterWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: ${ ({ margin }) => margin };
`;

export const RangeSliderInnerWrapper = styled.div`
  display: flex;
  height: ${ ({ height }) => height };
  position: relative;
  align-items: center;
`;

export const ValWrapper = styled.div`
  margin: ${ ({ margin }) => margin };
  color: ${ ({ theme }) => theme.font };
  display: flex;
  align-items: center;
  width: 4rem;
`;

const thumbPartial = css`
  height: 2rem;
  width: 2rem;
  border-radius: 50%;
  background: ${ ({ theme }) => theme.primary };
  cursor: pointer;
  margin-top: -.8rem;
  z-index: 10000;
`;

const trackPartial = css`
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: ${ ({ theme }) => theme.grey };
  border-radius: 1.3px;
`;

const focusPartial = css`
  
`;

export const Slider = styled.input`
  -webkit-appearance: none;
  width: 100%;
  background: transparent;
  &:focus {
    outline: none;
  }
  &::-ms-track {
    width: 100%;
    cursor: pointer;
    background: transparent;
    border-color: transparent;
    color: transparent;
  } 
  &::-webkit-slider-thumb {
    -webkit-appearance: none;
    ${ thumbPartial }
  }
  &::-moz-range-thumb {
    ${ thumbPartial }
  }
  &::-ms-thumb {
    ${ thumbPartial }
  }
  &::-webkit-slider-runnable-track {
    ${ trackPartial }
  }
  &:focus::-webkit-slider-runnable-track {
    ${ focusPartial }
  }
  &::-moz-range-track {
      ${ trackPartial }
  }
  &::-ms-track {
    width: 100%;
    height: .4rem;
    cursor: pointer;
    background: transparent;
    border-color: transparent;
    color: transparent;
  }
  &::-ms-fill-lower {
    ${ trackPartial }
  }
  &:focus::-ms-fill-lower {
    ${ trackPartial }
  }
  &::-ms-fill-upper {
    ${ trackPartial }
  }
  &:focus::-ms-fill-upper {
    ${ focusPartial }
  }
  &[disabled] {
    cursor: not-allowed;
    opacity: 0.75;
  }
`;

export const ValueDisplay = styled.div`
  background-color: ${ ({ theme }) => theme.grey };
  height: ${ ({ size }) => size };
  width: ${ ({ size }) => size };
  border-radius: .5rem;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 1.2rem;
  margin-left: 1rem;
`;