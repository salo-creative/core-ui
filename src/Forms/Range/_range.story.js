import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, number } from '@storybook/addon-knobs';
import { State, Store } from '@sambego/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Range } from '../../index';

// README
import README from './README.md';

const stories = storiesOf('Forms | Range', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['rangeslider'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: 50 });
    // KNOBS
    const label = text('Label', 'Range label');
    const error = boolean('Show error state', false);
    const errorMessage = text('Error Text', 'Something went wrong!');
    const helperText = text('Helper text', 'Helper text');
    const step = number('Step', 1);
    const disabled = boolean('Show disabled state', false);

    return (
      <State store={ store }>
        { state => (
          <Range
            disabled={ disabled }
            error={ error }
            errorMessage={ errorMessage }
            helperText={ helperText }
            label={ label }
            name='story'
            min={ 0 }
            max={ 100 }
            value={ state.value }
            onChange={ ({ value }) => {
              store.set({ value });
            } }
            step={ step }
          />
        ) }
      </State>
    );
  }),
  {
    info: { propTablesExclude: [State] },
    notes: README
  }
);