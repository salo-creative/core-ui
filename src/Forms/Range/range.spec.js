import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Range from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Range />
  );
  expect(container.firstChild).toMatchSnapshot();
});