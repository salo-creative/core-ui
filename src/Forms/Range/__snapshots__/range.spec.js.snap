// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 0 2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 4rem;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  -webkit-appearance: none;
  width: 100%;
  background: transparent;
}

.c2:focus {
  outline: none;
}

.c2::-ms-track {
  width: 100%;
  cursor: pointer;
  background: transparent;
  border-color: transparent;
  color: transparent;
}

.c2::-webkit-slider-thumb {
  -webkit-appearance: none;
  height: 2rem;
  width: 2rem;
  border-radius: 50%;
  background: #00ACED;
  cursor: pointer;
  margin-top: -.8rem;
  z-index: 5;
}

.c2::-moz-range-thumb {
  height: 2rem;
  width: 2rem;
  border-radius: 50%;
  background: #00ACED;
  cursor: pointer;
  margin-top: -.8rem;
  z-index: 5;
}

.c2::-ms-thumb {
  height: 2rem;
  width: 2rem;
  border-radius: 50%;
  background: #00ACED;
  cursor: pointer;
  margin-top: -.8rem;
  z-index: 5;
}

.c2::-webkit-slider-runnable-track {
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: #e9ecef;
  border-radius: 1.3px;
}

.c2::-moz-range-track {
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: #e9ecef;
  border-radius: 1.3px;
}

.c2::-ms-track {
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: transparent;
  border-color: transparent;
  color: transparent;
}

.c2::-ms-fill-lower {
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: #e9ecef;
  border-radius: 1.3px;
}

.c2:focus::-ms-fill-lower {
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: #e9ecef;
  border-radius: 1.3px;
}

.c2::-ms-fill-upper {
  width: 100%;
  height: .4rem;
  cursor: pointer;
  background: #e9ecef;
  border-radius: 1.3px;
}

.c2[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

.c3 {
  background-color: #e9ecef;
  height: 4rem;
  width: 4rem;
  border-radius: .5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 1.2rem;
  margin-left: 1rem;
}

<div
  class="c0"
>
  <div
    class="c1"
    height="4rem"
  >
    <input
      aria-invalid="false"
      aria-required="false"
      class="c2"
      max="100"
      min="0"
      type="range"
      value=""
    />
    <div
      class="c3"
    />
  </div>
</div>
`;
